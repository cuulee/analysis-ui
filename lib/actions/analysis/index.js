/**
 * @flow
 *
 * Redux actions related to Single Point Analysis requests.
 */

import lonlat from '@conveyal/lonlat'
import {fetchError, fetchMultiple} from '@conveyal/woonerf/fetch'
import {Map as LeafletMap} from 'leaflet'
import {createAction} from 'redux-actions'
import {TextDecoder} from 'text-encoding'

import {
  ANALYSIS_URL,
  PROFILE_REQUEST_DEFAULTS,
  TRAVEL_TIME_PERCENTILES
} from '../../constants'
import {
  COMPUTING_ISOCHRONE,
  INITIALIZING_CLUSTER,
  PERFORMING_ANALYSIS
} from '../../constants/analysis-status'
import {saveToServer as saveProject} from '../project'
import selectComparisonInProgress from '../../selectors/comparison-in-progress'
import selectCurrentProject from '../../selectors/current-project'
import selectCurrentProjectId from '../../selectors/current-project-id'
import timeout from '../../utils/timeout'

import R5Version from '../../modules/r5-version'

import type {GetState, LonLat} from '../../types'

const RETRY_TIMEOUT_MILLSECONDS = 5000
const SURFACE_HEADER_LENGTH = 9
const SURFACE_HEADER_TITLE = 'ACCESSGR'

export const clearIsochroneResults = createAction('clear isochrone results')
export const clearComparisonProject = createAction('clear comparison project')
export const enterAnalysisMode = createAction('enter analysis mode')
export const exitAnalysisMode = createAction('exit analysis mode')
export const setActiveVariant = createAction('set active variant')

export const setComparisonProject = createAction('set comparison project')
export const setDestination = createAction('set destination')
export const setIsochroneCutoff = createAction('set isochrone cutoff')
export const setIsochroneFetchStatus = createAction(
  'set isochrone fetch status'
)
export const setIsochroneFetchStatusMessage = createAction(
  'set isochrone fetch status message'
)
export const setIsochroneLonLat = createAction('set isochrone lonlat', x =>
  lonlat(x)
)
export const setIsochroneResults = createAction('set isochrone results')
export const setProjectApplicationErrors = createAction(
  'set project application errors'
)
export const setProjectApplicationWarnings = createAction(
  'set project application warnings'
)
export const selectBookmark = createAction('select bookmark')

const setTravelTimeSurface = createAction('set travel time surface')
const setComparisonTravelTimeSurface = createAction(
  'set comparison travel time surface'
)

/**
 * Save the profile request parameters to the project
 */
export const setProfileRequest = createAction('set profile request')

/**
 * Handle fetching and constructing the travel time surface and comparison
 * surface and dispatching updates along the way.
 */
export const fetchTravelTimeSurface = () =>
  (dispatch: Dispatch, getState: GetState) => {
    dispatch(setIsochroneFetchStatus(PERFORMING_ANALYSIS))

    const state = getState()
    const {lon, lat} = state.analysis.isochroneLonLat || state.mapState.center

    const project = selectCurrentProject(state)
    const currentProjectId = selectCurrentProjectId(state)
    const comparisonInProgress = selectComparisonInProgress(state)
    const workerVersion = R5Version.selectors.selectCurrentR5Version(state, {})

    const compareToStraightLineDistance = /compareToStraightLineDistance=([0-9.]+)/.exec(
      window.location.search
    )

    const profileRequest = {
      ...PROFILE_REQUEST_DEFAULTS,
      ...state.analysis.profileRequest,
      fromLat: lat,
      fromLon: lon,
      workerVersion,
      percentiles: TRAVEL_TIME_PERCENTILES,
      projectId: currentProjectId
    }

    // Store the profile request settings to the project
    dispatch(saveProject({
      ...project,
      analysisRequestSettings: profileRequest
    }))

    async function retry (response) {
      if (response.status === 202) {
        dispatch(setIsochroneFetchStatusMessage(INITIALIZING_CLUSTER))
        await timeout(RETRY_TIMEOUT_MILLSECONDS)
        return true
      } else {
        return false
      }
    }

    const fetches = [{
      url: ANALYSIS_URL,
      options: {
        method: 'post',
        body: profileRequest
      },
      retry
    }]

    if (comparisonInProgress && !compareToStraightLineDistance) {
      fetches.push({
        url: ANALYSIS_URL,
        options: {
          method: 'post',
          body: {
            ...profileRequest,
            projectId: state.analysis.comparisonProjectId, // override with comparison values
            variantIndex: state.analysis.comparisonVariant
          }
        },
        retry
      })
    }

    dispatch(fetchMultiple({
      fetches,
      next (error, responses) {
        if (error) {
          dispatch([
            setProjectApplicationWarnings(null),
            setTravelTimeSurface(null),
            setComparisonTravelTimeSurface(null),
            setIsochroneFetchStatus(false)
          ])

          // responses is just the single response when there was an error
          if (Array.isArray(responses.value)) {
            dispatch(setProjectApplicationErrors(responses.value))
          } else {
            dispatch(fetchError(responses.value))
          }
        } else {
          dispatch(setIsochroneFetchStatusMessage(COMPUTING_ISOCHRONE))
          const [surface, comparisonSurface] = responses.map(response =>
            responseToSurface(response.value))

          let straightLineSurface = null
          if (compareToStraightLineDistance) {
            const speedKmh = parseFloat(compareToStraightLineDistance[1])
            straightLineSurface = createStraightLineDistanceTravelTimeSurface(
              surface,
              {lon, lat},
              speedKmh
            )
          }

          return [
            setProjectApplicationErrors(null),
            setProjectApplicationWarnings([
              ...surface.warnings,
              ...(comparisonSurface ? comparisonSurface.warnings : [])
            ]),
            setTravelTimeSurface(surface),
            setComparisonTravelTimeSurface(straightLineSurface || comparisonSurface),
            setIsochroneFetchStatus(false),
            // in case multiple requests are enqueued simultaneously, attempt to keep the UI in sync
            setIsochroneLonLat({lon, lat}),
            setProfileRequest(profileRequest)
          ]
        }
      }
    }))
  }

// exported for testing
export function responseToSurface (response: any): any {
  if (response[0] && response[0].title) {
    // this is a list of errors from the backend
    return {
      errors: response,
      warnings: []
    }
  } else {
    // First read the header to figure out how big the binary portion is, then
    // read the full binary portion, then read the sidecar metadata at the end.
    const header = new Int32Array(response, 0, SURFACE_HEADER_LENGTH)
    // validate header and version
    if (intToString(header[0]) + intToString(header[1]) !== SURFACE_HEADER_TITLE) {
      throw new Error('Invalid header in travel time surface')
    }
    if (header[2] !== 0) {
      throw new Error(`Unsupported version ${header[2]} of travel time surface`)
    }
    const zoom = header[3]
    const west = header[4]
    const north = header[5]
    const width = header[6]
    const height = header[7]
    const nSamples = header[8]

    // 9 ints of header, each four bytes wide
    const data = new Int32Array(
      response,
      SURFACE_HEADER_LENGTH * 4,
      width * height * nSamples
    )

    // de delta code data
    for (let pixel = 0, y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        // pixels are delta coded individually
        for (let sample = 0, val = 0; sample < nSamples; sample++, pixel++) {
          data[pixel] = val += data[pixel]
        }
      }
    }

    // read metadata
    // TODO cross browser compatibility
    const decoder = new TextDecoder('utf-8') // utf-8 is Jackson default
    const rawMetadata = new Uint8Array(
      response,
      (SURFACE_HEADER_LENGTH + width * height * nSamples) * 4
    )
    const metadata = JSON.parse(decoder.decode(rawMetadata))

    return {
      zoom,
      west,
      north,
      width,
      height,
      nSamples,
      errors: [], // no errors - we got a result
      warnings: metadata.projectApplicationWarnings || [],
      get (x: number, y: number) {
        const index1d = (y * width + x) * nSamples
        return data.slice(index1d, index1d + nSamples)
      }
    }
  }
}

/**
 * Convert a four-byte int to a four-char string
 */
function intToString (val) {
  return (
    String.fromCharCode(val & 0xff) +
    String.fromCharCode((val >> 8) & 0xff) +
    String.fromCharCode((val >> 16) & 0xff) +
    String.fromCharCode((val >> 24) & 0xff)
  )
}

/**
 * Create a straight-line-distance travel time surface. This is a hidden feature
 * to allow a particular client to demonstrate that straight-line distance _from
 * the origin_ is absurd when used to compute transit access. Yes, people
 * actually do this.
 */
function createStraightLineDistanceTravelTimeSurface (surface: any, origin: LonLat, speedKmh: number) {
  const array = new Uint8ClampedArray(surface.nSamples)

  return {
    ...surface,
    errors: [],
    warnings: [{
      title: `Comparison project is using straight line distance with speed ${speedKmh} km/h`,
      messages: []
    }],
    get (x, y) {
      const latlng = LeafletMap.prototype.unproject([x + surface.west, y + surface.north], surface.zoom)
      const distMeters = latlng.distanceTo([origin.lat, origin.lon])
      const timeMinutes = (distMeters / 1000 / speedKmh * 60) | 0

      for (let i = 0; i < surface.nSamples; i++) {
        array[i] = timeMinutes
      }

      return array
    }
  }
}
