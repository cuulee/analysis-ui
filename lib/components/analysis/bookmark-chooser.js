// @flow
import Icon from '@conveyal/woonerf/components/icon'
import Pure from '@conveyal/woonerf/components/pure'
import sortBy from 'lodash/sortBy'
import React from 'react'
import Select from 'react-select'

import {Group} from '../input'
import {Button} from '../buttons'
import messages from '../../utils/messages'

import type {Bookmark} from '../../types'

type Props = {
  bookmarks: Bookmark[],
  bookmarkData: any,
  disabled: boolean,
  selectedBookmark: boolean,

  createBookmark: () => void,
  selectBookmark: (Bookmark) => void
}

export default class BookmarkChooser extends Pure {
  props: Props

  _selectBookmark = (e: {value: string}) => {
    const {bookmarks, selectBookmark} = this.props

    const bookmark = bookmarks.find(b => b._id === e.value)
    if (bookmark) selectBookmark(bookmark)
  }

  _createBookmark = () => {
    const {bookmarkData, bookmarks, createBookmark} = this.props
    const bookmarkName = window.prompt(
      'Enter a name for your bookmark',
      `Bookmark ${bookmarks.length + 1}`
    )

    if (bookmarkName && bookmarkName.length > 0) {
      createBookmark({
        ...bookmarkData,
        name: bookmarkName
      })
    }
  }

  render () {
    const {bookmarks, disabled, selectedBookmark} = this.props

    const options = sortBy(bookmarks, ['name']).map(b => ({
      label: b.name,
      value: b._id
    }))

    return (
      <Group label={messages.analysis.bookmark}>
        <div className='row'>
          <div className='col-xs-6'>
            <Select
              clearable={false}
              options={options}
              disabled={disabled} // don't allow users to reselect bookmarks while their bookmark is loading
              value={selectedBookmark}
              onChange={this._selectBookmark}
            />
          </div>
          <div className='col-xs-6'>
            <Button
              block
              disabled={disabled}
              onClick={this._createBookmark}
              style='success'
            >
              <Icon type='plus' /> {messages.analysis.createBookmark}
            </Button>
          </div>
        </div>
      </Group>
    )
  }
}
