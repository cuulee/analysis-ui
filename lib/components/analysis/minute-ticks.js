// @flow
import React from 'react'
import {sprintf} from 'sprintf-js'

import messages from '../../utils/messages'

// The x axis labels for the plot
const TIME_LABELS = [15, 30, 45, 60, 75, 90, 105, 120]

type Props = {
  scale(number): number,
  textHeight: number
}

/**
 * Render the minute axis ticks - shared by stacked percentile curves plot and
 * travel time distribution plot
 */
export default function MinuteTicks ({scale, textHeight}: Props) {
  return (
    <g>
      {/* We don't explicitly center the '15 minutes' text; fudge the left side a bit so that the 15 appears centered */}
      {TIME_LABELS.map((v, i, arr) => (
        <text
          x={i === 0 ? scale(v) - textHeight / 2 : scale(v)}
          y={0}
          key={`x-tick-${v}`}
          style={{
            textAnchor: i === arr.length - 1
              ? 'end' // don't run off end, fudge position of last value a bit
              : i === 0
                ? 'start' // don't center the whole '15 minutes' text
                : 'middle'
          }}
        >
          {/* label first minute */}
          {i === 0 ? sprintf(messages.analysis.minutes, v) : v}
        </text>
      ))}
    </g>
  )
}
