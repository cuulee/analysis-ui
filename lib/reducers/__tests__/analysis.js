// @flow

import {handleActions} from 'redux-actions'

import * as analysis from '../analysis'
import {mockProjectApplicationError} from '../../utils/mock-data'

describe('reducers > analysis', () => {
  const rawReducer = handleActions(analysis.reducers, analysis.initialState)
  // wrap reducer in order to snapshot test the ever-changing date
  const reducer = (state, action) => {
    const outState = rawReducer(state, action)
    if (outState && outState.profileRequest && outState.profileRequest.date) {
      outState.profileRequest.date = '2016-12-20'
    }
    return outState
  }

  // Default State Test
  it('should handle default state', () => {
    expect(reducer(undefined, {type: 'blah', payload: {}})).toMatchSnapshot()
  })

  // Specific Handler Tests
  it('should handle set profile request', () => {
    const action = {
      type: 'set profile request',
      payload: {
        fromTime: 24321,
        toTime: 25220
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle clear isochrone results', () => {
    const action = {type: 'clear isochrone results', payload: {}}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone cutoff', () => {
    const action = {type: 'set isochrone cutoff', payload: 123}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone fetch status', () => {
    const action = {type: 'set isochrone fetch status', payload: true}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone lonlat', () => {
    const action = {type: 'set isochrone lonlat', payload: 123}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set project application errors', () => {
    const action = {
      type: 'set project application errors',
      payload: [mockProjectApplicationError]
    }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle enter analysis mode', () => {
    const action = {type: 'enter analysis mode'}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle exit analysis mode', () => {
    const action = {type: 'exit analysis mode'}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set active variant', () => {
    const action = {
      type: 'set active variant',
      payload: 21
    }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison in progress', () => {
    const action = {
      type: 'set comparison in progress',
      payload: true
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison project', () => {
    const action = {
      type: 'set comparison project',
      payload: {
        _id: 'COMPARISON PROJECT ID',
        bundleId: 'COMPARISON BUNDLE ID',
        variantIndex: 16
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison modifications', () => {
    const action = {
      type: 'set comparison modifications',
      payload: {
        _id: 'COMPARISON PROJECT ID',
        bundleId: 'COMPARISON BUNDLE ID',
        variantIndex: 16,
        modifications: ['MODIFICATIONS']
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set active regional analyses', () => {
    const action = {
      type: 'set active regional analyses',
      payload: {
        _id: 'project id',
        comparisonId: 'comparison id'
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set regional analysis grids', () => {
    const action = {
      type: 'set regional analysis grids',
      payload: {
        grid: {which: 'GRID'},
        comparisonGrid: {which: 'COMPARISON GRID'},
        differenceGrid: {which: 'DIFFERENCE_GRID'},
        probabilityGrid: {which: 'PROBABILITY GRID'}
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })
})
