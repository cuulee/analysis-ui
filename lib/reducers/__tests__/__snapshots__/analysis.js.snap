// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers > analysis should handle clear isochrone results 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochrone": null,
  "isochroneLonLat": null,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle default state 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle enter analysis mode 1`] = `
Object {
  "active": true,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle exit analysis mode 1`] = `
Object {
  "active": false,
  "comparisonTravelTimeSurface": null,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "projectApplicationWarnings": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
  "travelTimeSurface": null,
}
`;

exports[`reducers > analysis should handle set active regional analyses 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": "project id",
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": "comparison id",
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set active variant 1`] = `
Object {
  "active": false,
  "activeVariant": 21,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set comparison in progress 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set comparison modifications 1`] = `
Object {
  "active": false,
  "comparisonBundleId": "COMPARISON BUNDLE ID",
  "comparisonModifications": Array [
    "MODIFICATIONS",
  ],
  "comparisonProjectId": "COMPARISON PROJECT ID",
  "comparisonVariant": 16,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set comparison project 1`] = `
Object {
  "active": false,
  "comparisonBundleId": "COMPARISON BUNDLE ID",
  "comparisonModifications": null,
  "comparisonProjectId": "COMPARISON PROJECT ID",
  "comparisonVariant": 16,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set isochrone cutoff 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set isochrone fetch status 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": true,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set isochrone lonlat 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneLonLat": 123,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set profile request 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 24321,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 25220,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set project application errors 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": Array [
    Object {
      "messages": Array [
        "Headway is not greater than zero.",
      ],
      "modificationId": "Add Trip Pattern",
      "title": "Error while applying the modification entitled 'Add Trip Pattern'.",
    },
  ],
  "regional": Object {
    "_id": null,
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "selectedBookmark": null,
}
`;

exports[`reducers > analysis should handle set regional analysis grids 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSpeed": 4.166666666666667,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxRides": 4,
    "maxTripDurationMinutes": 60,
    "monteCarloDraws": 200,
    "percentiles": Array [
      5,
      25,
      50,
      75,
      95,
    ],
    "toTime": 32400,
    "transitModes": "BUS,TRAM,RAIL,SUBWAY",
    "travelTimePercentile": 50,
    "variantIndex": 0,
    "walkSpeed": 1.3888888888888888,
  },
  "projectApplicationErrors": null,
  "regional": Object {
    "_id": null,
    "breaks": undefined,
    "comparisonGrid": Object {
      "which": "COMPARISON GRID",
    },
    "comparisonId": null,
    "differenceGrid": Object {
      "which": "DIFFERENCE_GRID",
    },
    "grid": Object {
      "which": "GRID",
    },
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": Object {
      "which": "PROBABILITY GRID",
    },
  },
  "selectedBookmark": null,
}
`;
